virttest/libvirt_xml/vm_xml.py

```diff
@@ -934,6 +951,40 @@ def get_disk_all(self):
             disks[dev] = node
         return disks
 
+    def get_disk_all_by_attr(self, reverse=False, **kwargs):
+        """
+        Return VM's disk from XML definition by attribute, None if not set
+        attribute or not found disks.
+        If mutiple attributes are specified, the disk will be returned only
+        when all attributes are matched.
+
+        Usage example:
+        1. get_disk_all_by_attr(device='disk')
+        2. get_disk_all_by_attr(type='file', device='disk')
+        3. get_disk_all_by_attr(device='cdrom', reverse=True)
+
+        :param reverse: If true, return disks that are not equal the specified
+            type or device attribute.
+        :param kwargs: attribute and value pairs of the disk
+        """
+        disk_nodes = self.xmltreefile.find('devices').findall('disk')
+        disks = {}
+        for node in disk_nodes:
+            matched = False
+            for attr_name, attr_val in kwargs.items():
+                if attr_val and eval('node.get(attr_name) %s attr_val' %
```

Comment:
I a little wonder how here handle the situation (type='file', device='disk', reverse=True).  More general case is that we need disk with device !=disk but type=file

Reply:
This function doesn't limit to specific attribute, and even when libvirt added more attributes, it still don't need to do any changes to adapt the new attribute.
Users only need to refer libvirt.org to get the disk attribute, and specifiy them as dictionary key-value pairs. this function will only return the result when all attributes are matched.

Reply:
Sorry, I misunderstood your question. 
You cannot do it by this function for device !=disk but type=file directly, in order to do that, you could select them separately and then do intersection operation by yourselves.
It's not the requirement in this function. may be a new function for that.

Reply:
A simple implementation is make the dictionary be accepting like 'device':('disk', '!=') and 'type':('file', '==').

Reply:
A new function could be like get_disk_all_by_expr(self, **args).
And the usage could be get_disk_all_by_expr('device !=disk', 'type=file', ...)